### PR TITLE
opcode `DIV` and `MOD`

### PR DESCRIPTION
**Pull request by icemelon**
Spec: https://github.com/appliedzkp/zkevm-specs/pull/149

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/execution/mul.rs:58]: seoncd -> second and divosr -> divisor

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:699]: maxium -> maximum

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/execution/mul.rs:85]: there is a `require_zero` method you can use here that is more clear imo. all it does is call `add_constraint`.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:687]: opcdoe MUL -> opcode MUL

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:713]: Can you append "in the MulWordsGadget", since we don't actually check do that check in this gadget?

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/execution/mul.rs:178]: Can you fix the unclosed parenthesis in the comments in line 170?

**Review by CPerezz — COMMENTED** [review]
So far the impl idea seems correct to me. 
Anyway, I think it should be better to wait until the PR to the specs is done to merge this.

Just left a couple of nits/questions :smile:

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/mul.rs]: Fixed

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/mul.rs:177]: Fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs]: Updated the doc

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs]: Fixed

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/mul.rs]: Updated

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs]: Fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs]: it's removed

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util.rs:286]: added the doc

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/mul.rs:76]: I think we can still use the expression to derive `is_mul` unless the degree of constraints is too high. What do you think?

**Review by CPerezz — APPROVED** [review]
LGTM!

Nice job @icemelon  :+1: 


Edit: I thought the `is_mul/div/mod` was addressed. Excuse me. Let's check it before merging and make sure if it's needed or not.

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/mul.rs:76]: But even if you derive it, wouldn't it be possible that it is not `0` or `1`? 
I mean, you're kinda using it as a selector, and so, it would make sense that we constrain it to be within `[0,1]`. 
Maybe this is done earlier or something similar. In which case, ignore this comment/discussion.

Otherways, I think we need to constrain it. Could you explain how deriving from the expression will guarantee that the value is within the correct range?

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/mul.rs:80]: Same concern as explained above.

Maybe here you can save it since you do `1.expr() - is_mul` which kinda goes in that way. But I still wonder if it's needed or not.

**Review by CPerezz — COMMENTED** [review]
Just the question about the binary constraint for the `is_mul/div/mod`.

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/mul.rs:76]: `is_mul`, `is_div` and `is_mod` can only be 0 or 1. This is because we have constrained the `opcode` to be chosen from the set `(MUL, DIV, MOD)` through the [`ResponsibleOpcode` lookup](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/util/common_gadget.rs#L37-L47).

**Comment by icemelon** [comment]
@CPerezz The spec is up at https://github.com/appliedzkp/zkevm-specs/pull/149

File: zkevm-circuits/src/evm_circuit/execution/mul.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/mul.rs:76]: Thanks! That clarifies the point completely!

File: zkevm-circuits/src/evm_circuit/witness.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/witness.rs:1077]:
```suggestion
                    OpcodeId::MUL | OpcodeId::DIV | OpcodeId::MOD => ExecutionState::MUL_DIV_MOD,
```

Maybe this to fit in with the rest of naming?

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:721]: Can we consider to rename this gadget so that it reflects all of the Math opcodes it can be used for?

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:825]: You should be able to use `try_collect`. But use what is easier for you. Just leaving as suggestion.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:831]: ditto

**Review by CPerezz — APPROVED** [review]
LGTM. Left some suggestions but they're not mandatory at all! :)

**Comment by CPerezz** [comment]
@icemelon should we update this branch and merge?

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs:825]: `Map` doesn't provide `try_collect`.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs:831]: ditto

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs:721]: I think the current gadget name reflects what it does that just multiple two words and add an additional word. So if rename it to sth like `MulDivModGadget`, it will clash with `MulDivModGadget` for opcode gadget

File: zkevm-circuits/src/evm_circuit/witness.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/witness.rs:1077]: I remember I had a similar suggestion for `ADD` -> `ADD_SUB`. @han0110 thought it's a bit confusing. So what do you think about the execution state renaming?

**Comment by icemelon** [comment]
@CPerezz Updated to latest main now.

**Comment by lispc** [comment]
I wish github supported three way merge on webpage... So we don't need to run `git mergetool` locally so often..

File: zkevm-circuits/src/evm_circuit/witness.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/witness.rs:1077]: If it's more clear in this way, we should just rename it. My thought was to not create new term, but it seems inevitable.

Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/witness.rs:1077]: Sure. I'll make the change.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs:721]: That makes sense!

**Comment by icemelon** [comment]
@han0110 @ChihChengLiang @ed255 Could you help review this PR?

File: zkevm-circuits/src/evm_circuit/execution/mul_div_mod.rs
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/execution/mul_div_mod.rs:58]: divsor -> divisor

**Review by miha-stopar — APPROVED** [review]
LGTM!

**Comment by CPerezz** [comment]
@icemelon I think that with @miha-stopar 's review we can now merge once the conflicts are resolved! :smile:

File: zkevm-circuits/src/evm_circuit/execution/mul_div_mod.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/execution/mul_div_mod.rs:36]: nitpick 
```suggestion
    const NAME: &'static str = "MUL_DIV_MOD";
```

**Review by ChihChengLiang — APPROVED** [review]
LGTM

File: zkevm-circuits/src/evm_circuit/execution/mul_div_mod.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/mul_div_mod.rs]: Fixed

**Pull request merged**
